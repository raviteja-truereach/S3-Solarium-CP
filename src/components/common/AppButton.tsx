/**
 * App Button Component
 * Themed button extending react-native-paper Button
 */
import React from 'react';
import { Button, useTheme } from 'react-native-paper';
import type { ButtonProps } from 'react-native-paper';
import { StyleSheet } from 'react-native';

export interface AppButtonProps extends Omit<ButtonProps, 'children'> {
  /** Button text */
  title: string;
  /** Button variant */
  variant?: 'contained' | 'outlined' | 'text';
  /** Full width button */
  fullWidth?: boolean;
  /** Button size */
  size?: 'small' | 'medium' | 'large';
}

/**
 * AppButton Component
 * Consistent themed button with project styling
 */
export const AppButton: React.FC<AppButtonProps> = ({
  title,
  variant = 'contained',
  fullWidth = false,
  size = 'medium',
  style,
  contentStyle,
  ...props
}) => {
  const theme = useTheme();

  const buttonStyle = [
    styles.button,
    size === 'small' && styles.smallButton,
    size === 'large' && styles.largeButton,
    fullWidth && styles.fullWidth,
    style,
  ];

  const buttonContentStyle = [
    styles.content,
    size === 'small' && styles.smallContent,
    size === 'large' && styles.largeContent,
    contentStyle,
  ];

  return (
    <Button
      mode={variant}
      style={buttonStyle}
      contentStyle={buttonContentStyle}
      {...props}
    >
      {title}
    </Button>
  );
};

const styles = StyleSheet.create({
  button: {
    borderRadius: 8,
    marginVertical: 4,
  },
  content: {
    paddingVertical: 8,
    paddingHorizontal: 16,
  },
  smallButton: {
    borderRadius: 6,
  },
  smallContent: {
    paddingVertical: 6,
    paddingHorizontal: 12,
  },
  largeButton: {
    borderRadius: 12,
  },
  largeContent: {
    paddingVertical: 12,
    paddingHorizontal: 24,
  },
  fullWidth: {
    width: '100%',
  },
});

export default AppButton;
