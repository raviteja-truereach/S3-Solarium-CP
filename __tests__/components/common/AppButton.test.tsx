/**
 * AppButton Component Tests
 * Tests disabled state, accessibility, and backward compatibility
 */

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';
import { PaperProvider } from 'react-native-paper';
import { axe, toHaveNoViolations } from 'jest-axe';
import AppButton from '../../../src/components/common/AppButton';

expect.extend(toHaveNoViolations);

const TestWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
  <PaperProvider>{children}</PaperProvider>
);

describe('AppButton', () => {
  describe('Basic Functionality', () => {
    it('renders button with text correctly', () => {
      render(
        <TestWrapper>
          <AppButton testID="test-button">Click Me</AppButton>
        </TestWrapper>
      );

      expect(screen.getByText('Click Me')).toBeTruthy();
      expect(screen.getByTestId('test-button')).toBeTruthy();
    });

    it('calls onPress when button is pressed', () => {
      const mockPress = jest.fn();

      render(
        <TestWrapper>
          <AppButton onPress={mockPress} testID="test-button">
            Press Me
          </AppButton>
        </TestWrapper>
      );

      fireEvent.press(screen.getByTestId('test-button'));
      expect(mockPress).toHaveBeenCalledTimes(1);
    });

    it('calls onLongPress when button is long pressed', () => {
      const mockLongPress = jest.fn();

      render(
        <TestWrapper>
          <AppButton onLongPress={mockLongPress} testID="test-button">
            Long Press Me
          </AppButton>
        </TestWrapper>
      );

      fireEvent(screen.getByTestId('test-button'), 'onLongPress');
      expect(mockLongPress).toHaveBeenCalledTimes(1);
    });
  });

  describe('Disabled State', () => {
    it('does not call onPress when disabled', () => {
      const mockPress = jest.fn();

      render(
        <TestWrapper>
          <AppButton
            onPress={mockPress}
            disabled={true}
            testID="disabled-button"
          >
            Disabled Button
          </AppButton>
        </TestWrapper>
      );

      fireEvent.press(screen.getByTestId('disabled-button'));
      expect(mockPress).not.toHaveBeenCalled();
    });

    it('does not call onLongPress when disabled', () => {
      const mockLongPress = jest.fn();

      render(
        <TestWrapper>
          <AppButton
            onLongPress={mockLongPress}
            disabled={true}
            testID="disabled-button"
          >
            Disabled Button
          </AppButton>
        </TestWrapper>
      );

      fireEvent(screen.getByTestId('disabled-button'), 'onLongPress');
      expect(mockLongPress).not.toHaveBeenCalled();
    });

    it('has correct accessibility state when disabled', () => {
      render(
        <TestWrapper>
          <AppButton disabled={true} testID="disabled-button">
            Disabled Button
          </AppButton>
        </TestWrapper>
      );

      const button = screen.getByTestId('disabled-button');
      expect(button.props.accessibilityState).toEqual({ disabled: true });
      expect(button.props.accessibilityLabel).toBe(
        'Disabled Button – disabled'
      );
    });

    it('has correct accessibility label when disabled with custom label', () => {
      render(
        <TestWrapper>
          <AppButton
            disabled={true}
            accessibilityLabel="Add Lead"
            testID="disabled-button"
          >
            +
          </AppButton>
        </TestWrapper>
      );

      const button = screen.getByTestId('disabled-button');
      expect(button.props.accessibilityLabel).toBe('Add Lead – disabled');
    });
  });

  describe('Variants and Sizes', () => {
    it('renders different variants correctly', () => {
      const { rerender } = render(
        <TestWrapper>
          <AppButton variant="primary" testID="primary-button">
            Primary
          </AppButton>
        </TestWrapper>
      );

      expect(screen.getByTestId('primary-button')).toBeTruthy();

      rerender(
        <TestWrapper>
          <AppButton variant="secondary" testID="secondary-button">
            Secondary
          </AppButton>
        </TestWrapper>
      );

      expect(screen.getByTestId('secondary-button')).toBeTruthy();

      rerender(
        <TestWrapper>
          <AppButton variant="outline" testID="outline-button">
            Outline
          </AppButton>
        </TestWrapper>
      );

      expect(screen.getByTestId('outline-button')).toBeTruthy();
    });

    it('renders different sizes correctly', () => {
      const { rerender } = render(
        <TestWrapper>
          <AppButton size="small" testID="small-button">
            Small
          </AppButton>
        </TestWrapper>
      );

      expect(screen.getByTestId('small-button')).toBeTruthy();

      rerender(
        <TestWrapper>
          <AppButton size="large" testID="large-button">
            Large
          </AppButton>
        </TestWrapper>
      );

      expect(screen.getByTestId('large-button')).toBeTruthy();
    });
  });

  describe('Backward Compatibility', () => {
    it('works without disabled prop (default false)', () => {
      const mockPress = jest.fn();

      render(
        <TestWrapper>
          <AppButton onPress={mockPress} testID="default-button">
            Default Button
          </AppButton>
        </TestWrapper>
      );

      fireEvent.press(screen.getByTestId('default-button'));
      expect(mockPress).toHaveBeenCalledTimes(1);
    });

    it('works with all existing props', () => {
      const mockPress = jest.fn();

      render(
        <TestWrapper>
          <AppButton
            onPress={mockPress}
            variant="primary"
            size="medium"
            style={{ margin: 10 }}
            testID="full-props-button"
          >
            Full Props Button
          </AppButton>
        </TestWrapper>
      );

      expect(screen.getByTestId('full-props-button')).toBeTruthy();
      fireEvent.press(screen.getByTestId('full-props-button'));
      expect(mockPress).toHaveBeenCalledTimes(1);
    });
  });

  describe('Accessibility', () => {
    it('should not have accessibility violations', async () => {
      const { container } = render(
        <TestWrapper>
          <AppButton testID="accessibility-button">Accessible Button</AppButton>
        </TestWrapper>
      );

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should not have accessibility violations when disabled', async () => {
      const { container } = render(
        <TestWrapper>
          <AppButton disabled={true} testID="disabled-accessibility-button">
            Disabled Accessible Button
          </AppButton>
        </TestWrapper>
      );

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
