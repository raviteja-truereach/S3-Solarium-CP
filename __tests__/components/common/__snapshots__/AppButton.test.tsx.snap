// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppButton Basic Rendering matches snapshot with all variants: contained-button 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#007AFF",
          "borderRadius": 8,
          "marginVertical": 4,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#007AFF",
            "borderColor": "transparent",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 0,
            "flex": undefined,
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 8,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                [
                  {
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  },
                  false,
                  false,
                  undefined,
                ],
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#1C1B1F",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      undefined,
                      false,
                      {
                        "color": "#FFFFFF",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Contained
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`AppButton Basic Rendering matches snapshot with all variants: outlined-button 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 8,
          "marginVertical": 4,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "#E0E0E0",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flex": undefined,
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 8,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                [
                  {
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  },
                  false,
                  false,
                  undefined,
                ],
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#1C1B1F",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      undefined,
                      false,
                      {
                        "color": "#007AFF",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Outlined
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`AppButton Basic Rendering matches snapshot with all variants: text-button 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 8,
          "marginVertical": 4,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 0,
            "flex": undefined,
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 8,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                [
                  {
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  },
                  false,
                  false,
                  undefined,
                ],
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#1C1B1F",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 12,
                      },
                      undefined,
                      false,
                      {
                        "color": "#007AFF",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Text
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`AppButton Basic Rendering matches snapshot with default props 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#007AFF",
          "borderRadius": 8,
          "marginVertical": 4,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#007AFF",
            "borderColor": "transparent",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 0,
            "flex": undefined,
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 8,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                [
                  {
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  },
                  false,
                  false,
                  undefined,
                ],
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#1C1B1F",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      undefined,
                      false,
                      {
                        "color": "#FFFFFF",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Default Button
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
